### PR TITLE
IGNITE-22732 Transaction aware SQL

### PR DESCRIPTION
**Pull request by nizhikov**
This PR includes 3 tickets:
[IGNITE-22732](https://issues.apache.org/jira/browse/IGNITE-22732) [Transaction aware SQL](https://github.com/apache/ignite/commit/3a124f69ba37d0fd5cf26fd3230c12d6103a9030)
[IGNITE-22874](https://issues.apache.org/jira/browse/IGNITE-22874) Support Transaction aware SQL in thin client
[IGNITE-23090](https://issues.apache.org/jira/browse/IGNITE-23090) [Support SQL transactions in thin JDBC](https://github.com/apache/ignite/commit/8a4bc8a5bcd70ab235fd84ed3914790e78a0f05d)

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by nizhikov** [comment]
Three PR's was reviewed by @alex-plekhanov  and @timoninmaxim 
I will get bot visa and merge full PR in master.

So, If there are any objections, please, share.

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java [hunk 1]
```diff
@@ -103,8 +108,11 @@ public class JdbcThinTcpIo {
     /** Version 2.13.0. */
     private static final ClientListenerProtocolVersion VER_2_13_0 = ClientListenerProtocolVersion.create(2, 13, 0);
 
+    /** Version 2.14.0. */
+    private static final ClientListenerProtocolVersion VER_2_14_0 = ClientListenerProtocolVersion.create(2, 14, 0);
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java:112]: VER_2_17_0?

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java [hunk 2]
```diff
@@ -319,6 +335,13 @@ private HandshakeResult handshake(ClientListenerProtocolVersion ver) throws IOEx
         if (ver.compareTo(VER_2_13_0) >= 0)
             writer.writeString(connProps.getQueryEngine());
 
+        if (ver.compareTo(VER_2_14_0) >= 0) {
+            writer.writeByte(connProps.getTransactionConcurrency().ordinal());
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java:339]: Why not reuse JdbcSetTxParametersRequest message (and send it only when TX_AWARE_QUERIES is enabled)  without changing handshake and protocol version?

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcThinTransactionalSelfTest.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        }
+    }
+
+    /** */
+    @Test
+    public void testNoTxInNoTxIsolation() throws Exception {
+        try (Connection conn = DriverManager.getConnection(URL)) {
+            conn.setTransactionIsolation(TRANSACTION_NONE);
+
+            ResultSet rs = conn.prepareStatement("SELECT 1").executeQuery();
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/jdbc/JdbcThinTransactionalSelfTest.java:181]: fixed

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDmlIntegrationTest.java [hunk 4]
```diff
@@ -153,12 +150,7 @@ public void testInsertPrimitiveKey() {
 
         QueryEngine engine = Commons.lookupComponent(grid(1).context(), QueryEngine.class);
```
[the remark this answers is not in the document]
Comment (nizhikov) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDmlIntegrationTest.java:151]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java (hunk 1 above)
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java (hunk 2 above)
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinTcpIo.java:339]: Add of transaction parameters to handshake was requested by @timoninmaxim to reduce network messages.

**Review by alex-plekhanov — APPROVED** [review]
I've approved this PR, but:
1. I think txId in JDBC request/response is redundant. There are no cases now, where it can helpful, but it leads to protocol and code complication.
2. I think JDBC tx parameters shouldn't be passed in handshake, since we have dedicated message for this. Again, it leads to protocol and code complication.

**Pull request merged**